design/plugin-backup-and-restore-progress-design.md

```diff
@@ -0,0 +1,434 @@
+	// +optional
+	// +nullable
+	StartTimestamp *metav1.Time `json:"startTimestamp,omitempty"`
+
+	// CompletionTimestamp records the time a backup was completed.
+	// Completion time is recorded even on failed backups.
+	// Completion time is recorded before uploading the backup object.
+	// The server's time is used for CompletionTimestamps
+	// +optional
+	// +nullable
+	CompletionTimestamp *metav1.Time `json:"completionTimestamp,omitempty"`
+
+	// Progress holds the total number of bytes of the volume and the current
+	// number of backed up bytes. This can be used to display progress information
+	// about the backup operation.
+	// +optional
+	Progress VolumeOperationProgress `json:"progress,omitempty"`
+}
+
+type VolumeOperationProgress struct {
+	TotalBytes int64
+	BytesDone int64
+}
+
+type VolumePluginBackupSpec volume.SnapshotSpec
+type VolumePluginBackupStatus volume.SnapshotStatus
+
+type VolumePluginBackup struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec VolumePluginBackupSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status VolumePluginBackupStatus `json:"status,omitempty"`
+}
+```
+
+Notice that VolumePluginBackupSpec is same as volume.SnapshotSpec, and, VolumePluginBackupStatus is same as volume.SnapshotStatus
+
+For every backup operation of volume, Velero creates VolumePluginBackup CR before calling volumesnapshotter's CreateSnapshot API.
+
+In order to know the CR created for the particular backup of a volume, Velero adds following labels to CR:
+- `velero.io/backup-name` with value as Backup Name, and,
+- `velero.io/pv-name` with value as volume that is undergoing backup
+
+Backup name being unique won't cause issues like duplicates in identifying the CR.
+Labels will be set with the value returned from `GetValidName` function. (https://github.com/vmware-tanzu/velero/blob/master/pkg/label/label.go#L35).
+
+If Plugin supports showing progress of the operation it is performing, it does following:
+- finds the VolumePluginBackup CR related to this backup operation by using `tags` passed in CreateSnapshot call
+- updates the CR with the progress regularly.
+
+After return from `CreateSnapshot` in `takePVSnapshot`, currently Velero adds `volume.Snapshot` to `backupRequest`. Instead of this, CR will be added to `backupRequest`.
+During persistBackup call, this CR also will be backed up to backup location.
+
+In backupSyncController, it checks for any VolumePluginBackup CRs that need to be synced from backup location, and syncs them to cluster if needed.
+
+VolumePluginBackup will be useful as long as backed up data is available at backup location. When the Backup is deleted either by manually or due to expiry, VolumePluginBackup also can be deleted.
+
+`processRequest` of `backupDeletionController` will perform deletion of VolumePluginBackup before volumesnapshotter's DeleteSnapshot is called.
+
+#### Backward compatibility:
```

Comment:
It seems like we could potentially get rid of the `volume.Snapshot` type and not need to keep it around with the type-aliasing for the CRD. The existing `volumesnapshots.json.gz` files should still be able to be deserialized into the new `VolumePluginBackup` CRs, and then we can check for the existence of `metadata.name` to determine if it's a CR or not (per https://github.com/vmware-tanzu/velero/pull/2543/files#diff-a724dc7ff53f45b40b4c5a5f57e741dbR184). 

Does that sound right? I think it'd be simpler if we could avoid the type aliasing and the two structs, and just have the new API type.

Reply:
yes @skriss .. updated design accordingly